### PR TITLE
Restructure function inliner

### PR DESCRIPTION
**Pull request by gramalingam**
**Description**: 

Reimplement the function inliner. 

**Motivation and Context**
The current function inliner has several limitations and drawbacks.
(i) It doesn't handle sub-graphs (graph-valued attributes contained in control-flow nodes) in several contexts.
(ii) There is a duplication of renaming-logic, with renaming of tensors happening in two places.
(iii) The creation of a FunctionImpl as a temporary (from a FunctionProto), which is then copied over by duplicating all its nodes in the main graph.

The new implementation fixes these limitations.

### CODE REVIEW

File: onnxruntime/core/graph/function_utils.cc
Comment (xadupre) [inline onnxruntime/core/graph/function_utils.cc:321]: unsigned?

File: onnxruntime/core/graph/function_utils.cc
Comment (gramalingam) [reply onnxruntime/core/graph/function_utils.cc:292]: Not as far as I am aware of. This is the trick that has been used so far in ORT. ONNX itself does have a more robust mechanism of first collecting all identifiers to help create new identifiers. But I prefer that be done separately.

File: onnxruntime/core/graph/function_utils.cc
Comment (souptc) [inline onnxruntime/core/graph/function_utils.cc:172]: thanks for adding this subgraph support.

File: onnxruntime/core/graph/graph.cc
Comment (souptc) [inline onnxruntime/core/graph/graph.cc:566]: copy probably is fine as there is no initializers in function proto. but do we really need a copy?

File: onnxruntime/core/graph/graph.cc
Comment (souptc) [inline onnxruntime/core/graph/graph.cc:3997]: any reason why we prefer the NodeProto in the signature, instead of use Graph::Node? or, could we provide an overload with Graph::Node so we don't need convert to nodeproto here?

File: onnxruntime/core/graph/function_utils.cc
Comment (xadupre) [inline onnxruntime/core/graph/function_utils.cc:304]: remove?

File: onnxruntime/core/graph/graph.cc
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:3997]: I think we could promote/move the underlying code into ONNX eventually (purely proto-based code). Yes, I can add an overload with Graph::Node.

File: onnxruntime/core/graph/graph.cc
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc:566]: Good question. I thought about it earlier. But decided this was fine for current use, because we end up changing the names/attributes when inlining. So, we end up with a copy anyway. (The current implementation reduces the number of intermediate-representations used during inlining from two to one. Reducing the one to zero is possible, but would complicate the code, so settled on this.)

File: onnxruntime/test/framework/function_test.cc
Comment (gramalingam) [reply onnxruntime/test/framework/function_test.cc:128]: Good point. Added these test-cases.

File: onnxruntime/core/graph/graph.cc
Comment (gramalingam) [reply onnxruntime/core/graph/graph.cc]: Added the overload.

File: onnxruntime/core/graph/function_utils.cc
Comment (liqunfu) [inline onnxruntime/core/graph/function_utils.cc:272]: When debugging with onnx-script optional input, I am hitting symbol_table being nullptr exception. Not use the root cause if here and if this Pr solves the issue. I am urging test this code with onnx-script because now, I just merged with function-experiment with master branch and I see too many breaks :(

Comment (gramalingam) [reply onnxruntime/core/graph/function_utils.cc:272]: @liqunfu : this PR requires the update to ONNX version. I think that might be the problem for the issues?

**Review by souptc — APPROVED** [review]

**Comment by garymm** [comment]
This is needed to support ONNX 1.12, in particular for the new SequenceMap op, which is a function that uses subgraphs.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/OnnxMl.cs
Comment (xadupre) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/OnnxMl.cs:6147]: Is this related to your changes?

Comment (garymm) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/OnnxMl.cs:6147]: I accidentally merged this with the ONNX 1.12 change. Once I merge ONNX 1.12 we will rebase this on master and these changes should go away.

**Comment by garymm** [comment]
ONNX 1.12 has been merged. @gramalingam can you please rebase, test and merge?

**Review by garymm — DISMISSED** [review]

**Comment by gramalingam** [comment]
@garymm : can you approve again? The last update dismissed your previous approval, thanks!

**Review by garymm — APPROVED** [review]

**Pull request merged**
